File: cpp/206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
        if(head == NULL || head->next == NULL) return head;
        
        ListNode* p = NULL;
        ListNode* q = head;
        ListNode* r = head->next;
        
        while(q != NULL){
            
            q->next = p;
            p = q;
            q = r;
            if(r) r = r->next;
            
        }
        
        return p;
        
    }
};

File: py/206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
def solve(curr,prev,head):
    if curr is None:
        head=prev
        return head
    head=solve(curr.next,curr,head)
    
    curr.next=prev
    return head
class Solution(object):
    def reverseList(self, head):
        return solve(head,None,head)
        
